[Bookings] Error and loading states for Team member filter screen

irfano opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->
Closes WOOMOB-1464

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
This adds handling for loading and error states when fetching team members on the Team Member filter page.

### Test Steps
<!-- Describe how to test your changes. Include only what’s needed for the reviewer to validate the behavior.
For new features, outline the main user flow or goal rather than every tap or click — the reviewer should be able to complete the flow naturally.
If applicable, mention key devices, scenarios, or edge cases to verify. -->
1. Log in with a CIAB store.
2. Go to Bookings.
3. Go to the All tab.
4. Tap the Filters button.
5. Select Team member

#### Tip for creating "first load" case
Use `DELETE FROM BookingResources` command in "App Inspection → Database Inspector → New Query" tab.

#### Tip for the API error case
[Temporarily_hardcode_siteId_for_Bookings_resources_request.patch](https://github.com/user-attachments/files/23557260/Temporarily_hardcode_siteId_for_Bookings_resources_request.patch): Use this patch to break site parameter when fetching resource.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
[Loading.webm](https://github.com/user-attachments/assets/ad223bd7-77a5-448b-bf3d-617ab7243bc9)

[Error.webm](https://github.com/user-attachments/assets/789c1c0e-1559-4c9d-8ad4-6e3dd1ec2493)

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

irfano committed:
Add loading state and snackbar support to BookingTeamMemberFilterPage

irfano committed:
Handle team members loading state

irfano committed:
Add bookings resources fetch error string

irfano committed:
Add loading state and error handling tests to BookingTeamMemberFilterViewModel

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/filter/teammember/BookingTeamMemberFilterViewModel.kt:
```diff
@@ -30,12 +32,26 @@ class BookingTeamMemberFilterViewModel @AssistedInject constructor(
             bookingsRepository.observeResources().distinctUntilChanged().collect { resources ->
-                _uiState.update { current -> current.copy(teamMembers = listOf(TeamMember.any) + resources) }
+                _uiState.update { current ->
+                    current.copy(isLoading = resources.isEmpty(), teamMembers = listOf(TeamMember.any) + resources)
```
I had to use `isLoading = resources.isEmpty()` here because `observeResources()` returns empty resources when the database is empty.

AdamGrzybkowski replied:
I think it would be better to set the `isLoading` to true below when fetching the resources and have an extra check whether the list is empty or not. The `isLoading=true` is quite hidden now and doesn't reflect what's happening, because the fetch happens even if the database result is not empty.

AdamGrzybkowski replied:
In fact, the way it is now causes a bug. If you enter the filter screen without a network connection and the booking resources table is empty, the loading skeleton won't go away. 


https://github.com/user-attachments/assets/a9f2ed54-62fc-48d0-b9e0-913268897651

irfano replied:
> The isLoading=true is quite hidden now

That's right, I agree. 

> If you enter the filter screen without a network connection and the booking resources table is empty, the loading skeleton won't go away.

I actually did it intentionally since we don’t have a dedicated empty or error screen. But as you mentioned, showing the loading state for empty cases could be misleading. I’ll update it.

irfano replied:
> In fact, the way it is now causes a bug. If you enter the filter screen without a network connection and the booking resources table is empty, the loading skeleton won't go away.

Fixed: c5e79b4f3fb602e9f18642101f3e6effef00c1a7

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/filter/teammember/BookingTeamMemberFilterPage.kt:
```diff
@@ -17,10 +34,65 @@ fun BookingTeamMemberFilterRoute(
+    Column(modifier = Modifier.fillMaxWidth()) {
+        SkeletonView(
+            modifier = Modifier
+                .size(62.dp, 64.dp)
```
I took these sizes from the sample screen in Figma.

AdamGrzybkowski commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/filter/teammember/BookingTeamMemberFilterPage.kt:
```diff
@@ -17,10 +34,65 @@ fun BookingTeamMemberFilterRoute(
         }
     val uiState by viewModel.uiState.collectAsState()
-    BookingTeamMemberFilterPage(uiState)
+    val event by viewModel.event.observeAsState()
```
FYI

About the `val uiState by viewModel.uiState.collectAsState()`

I had a long discussion with Hicham about `StateFlow` vs `LiveData` and the general consensus was that we should stick the `LiveData` - [link](https://github.com/woocommerce/woocommerce-android/pull/14620#discussion_r2354609417)


I know we missed this previously, but I think we should stick to the above rule for all filter screens as well.  <-- This doesn't mean we should change it in this PR, just wanted to bring awareness.

irfano replied:
Good suggestion! I updated it: 3bb08cc0c8f57de80b2ac069705e94ee2e49b86b
I also updated it for other filter pages in this PR. 08f29db593c4c1e8013d1885f36cf35ade3c4455 93696c3091c8ee55693c29fc4ddd3395610491ba

AdamGrzybkowski commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/filter/teammember/BookingTeamMemberFilterPage.kt:
```diff
@@ -17,10 +34,65 @@ fun BookingTeamMemberFilterRoute(
         }
     val uiState by viewModel.uiState.collectAsState()
-    BookingTeamMemberFilterPage(uiState)
+    val event by viewModel.event.observeAsState()
```
⚠️ `viewModel.event.observeAsState()` will persist the latest event during config changes, so the snack will reappear when rotating the device. 

Here's an example how Hicham wanted to solve this with a DisposableEffect - https://github.com/woocommerce/woocommerce-android/pull/14834/commits/dc8f9c7efb2f8af96249fca13fde655f92972dd5

irfano replied:
I tried using `DisposableEffect`, but the behavior didn't change, because the actual reason is that we call `fetchResources()` in the `init` function. 

The snackbar message doesn't persist. As you can see in the video below, once the network issue is resolved and I rotate the device, the snackbar is not shown again, which is expected:

[Screen_recording_20251125_000157.webm](https://github.com/user-attachments/assets/651863e7-045e-4f0e-bc5f-f50afb9ac595)

Calling `fetchResources()` again from `init` isn't really a major issue. It's actually the same on Orders screen. I also tested the Calendar app and saw the same behavior there. So, I think keeping it as is is fine. WDYT?

[Screen_recording_20251125_000506.webm](https://github.com/user-attachments/assets/58ef204d-d3a1-4d94-87f0-f7a785248268)

AdamGrzybkowski replied:
>I tried using DisposableEffect, but the behavior didn't change, because the actual reason is that we call fetchResources() in the init function.

This tells me that something is off then. The ViewModel should survive config changes (orientation change) and thus the `init` block shouldn't be invoked.

irfano replied:
The issue was in `BookingFilterListScreen`. Since we're controlling navigation in Compose, navigation was being triggered again even though we were already on the same page. I fixed it here: 1bafb5c832caa1c3c9c384449496ce93c7537b71
This will also fix the issue on the other sub-filter pages as well.

AdamGrzybkowski replied:
Thanks! Now with that change, we can clearly see that the event is persisted. I've added two logs, one in the ViewModel init block and one where the snackbar is shown. This is the result:


https://github.com/user-attachments/assets/6526b6d4-5620-40ac-b1f7-47cbca3d5eb9

irfano replied:
Sorry, I forgot to include DisposableEffect` commit. 🤦🏻‍♂️ I've added it now: a258125332b4e25039f368f0e87447f083adf5ec

AdamGrzybkowski commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/filter/teammember/BookingTeamMemberFilterPage.kt:
```diff
@@ -17,10 +34,65 @@ fun BookingTeamMemberFilterRoute(
+                .size(62.dp, 64.dp)
+                .padding(horizontal = 16.dp, vertical = 24.dp)
+        )
+        HorizontalDivider(thickness = 0.5.dp)
```
The divider should have a startPadding of 16.dp as well.

irfano replied:
I didn’t add that intentionally, to stay consistent with the Orders screen. We also don’t use `startPadding` on the other filter pages.

AdamGrzybkowski replied:
hmm the Customer filter has such padding. I guess we can either go with the new design and update other screens or modify the Customer filter to stay consistent.

irfano replied:
I checked the Product filter screens and we’re using an inset divider there as well. So I think the ideal approach is to use inset dividers across all screens in the Bookings tab, matching the Figma design. I’d like to open a separate PR for this task.

AdamGrzybkowski replied:
Sure, thanks!

AdamGrzybkowski commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/BookingsRepository.kt:
```diff
@@ -88,12 +88,13 @@ class BookingsRepository @Inject constructor(
     }
 
-    suspend fun fetchResources(): Result<Unit> {
+    suspend fun fetchResources(): Result<Array<BookingResourceEntity>> {
```
It doesn't look like we use this result, do we need it?

irfano replied:
Good catch! I used it during development, then stopped using it but left by mistake. I’ve removed the result: aca5f436cef044ade4d002bc0f564a23fc11b8a2

AdamGrzybkowski reviewed (changes requested):
Requesting changes because of the infinite skeleton when the table is empty.

irfano committed:
Refactor `fetchResources` return type

The `fetchResources` method in `BookingsRepository` no longer returns the fetched resources directly.

irfano committed:
Merge branch 'trunk' into issue/WOOMOB-1464-team-member-filter-error-loading

irfano committed:
Fix loading state for team member filter

Avoid hiding the loading indicator if the initial database query returns an empty list while a server fetch is in progress.

irfano committed:
Refactor `fetchResources` return type in tests

The return type of `bookingsRepository.fetchResources()` has changed from `Result<Array>` to `Result<Unit>`.

irfano committed:
Migrate BookingTypeFilter from StateFlow to LiveData

Use `observeAsState` to collect UI state from the ViewModel.

irfano committed:
Refactor BookingAttendanceStatusFilter to use LiveData

Migrate `uiState` from `StateFlow` to `LiveData` in `BookingAttendanceStatusFilterViewModel`.

irfano committed:
Convert `uiState` from `StateFlow` to `LiveData`

Converts `uiState` from a `StateFlow` to a `LiveData` in `BookingTeamMemberFilterViewModel` and updates the usage in `BookingTeamMemberFilterPage` accordingly.

AdamGrzybkowski commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/bookings/filter/teammember/BookingTeamMemberFilterViewModel.kt:
```diff
@@ -27,15 +30,32 @@ class BookingTeamMemberFilterViewModel @AssistedInject constructor(
+                var newUiStateValue = _uiState.value.copy(teamMembers = listOf(TeamMember.any) + resources)
+                if (resources.isNotEmpty()) {
+                    newUiStateValue = newUiStateValue.copy(isLoading = false)
+                }
```
I think we can improve this slightly. There's a potential timing issue here, and this place shouldn't really control the `isLoading` state. I would suggest adding this property to the UiState `val skeletonVisible: Boolean = isLoading && items.isEmpty()` and deleting these three lines. This way we will only show the skeleton when the fetch is happening and we don't have any items to show.

irfano replied:
Let me explain why I added the `isLoading` check into `observeResources()`. `observeResources()` returns data before `fetchResources()` completes, because it serves data from the cache. So when `observeResources()` returns non-empty data, we set `isLoading = false`.
If we don't update `isLoading` here, we have to wait for `fetchResources()` to finish before hiding skeleton view. I tried your suggestion, and it caused skeleton view to bi displayed even though we already had cached data.

AdamGrzybkowski replied:
This is true; the above wasn't enough. I missed that the `Any` is part of the list, so this would obviously fail. Here's the full patch [team_member_state_update.patch](https://github.com/user-attachments/files/23771136/team_member_state_update.patch) and how it works:

https://github.com/user-attachments/assets/cc826784-6a07-4892-ba7a-4f206259edbf

This creates a single point for the Skeleton visibility logic.

irfano replied:
Okay, I see. This just moves the `resources.isNotEmpty()` logic from the view model into the UI state without changing the behavior. It does add an extra property to the ui state model, though. Previously we only had `isLoading` to control the skeleton, and now we also expose the network call's loading state. We're not using this new property in the UI yet, but it's still a good improvement. I added it: 458995c6e1e5e49fe03ac0a99ef3bb5574ffece8

AdamGrzybkowski reviewed (commented):
Looks good, but I think we can improve the state a bit and we have to fix the ViewModel not surviving the config changes.

irfano committed:
Fix double navigation in booking filters

Prevent navigating to the same screen again if the current route already matches the destination.

irfano committed:
Fix `BookingTeamMemberFilterViewModelTest`

Ensure `uiState` is observed before running tests.

irfano committed:
Refactor snackbar handling in team member filter screen

Extract event handling logic into a separate `HandleEvents` composable.

irfano committed:
Update loading state handling in team member filter to use skeleton visibility

AdamGrzybkowski reviewed (approved):
Thanks for all the changes 🔥
